### DISCUSSION TITLE
[HELP]多进程调用下各子进程进度的显示

### DISCUSSION THREAD

**Discussion by xueli12** [RikaCelery/N_m3u8DL-RE#12] [Q&A]
在不开启 --force-ansi-console时，会显示各个进程的进度信息到最后一行，但只有百分比。日志文件中看不到进度条。
![image](https://github.com/RikaCelery/N_m3u8DL-RE/assets/51438016/69dbc340-7442-49f6-8e4a-8fe527399653)

开启了force-ansi-console后，会出现各个进程的窗口混在一起的情况，日志文件中也还是看不到进度条。
![image](https://github.com/RikaCelery/N_m3u8DL-RE/assets/51438016/8ebcf61f-83f0-4563-8958-391a92c359ec)

**Comment by RikaCelery** [comment]
啊首先，日志文件无论如何也不会出现进度条信息，日志文件只会写如下载的关键信息，并不是将屏幕内容原样保存到文件

**Comment by RikaCelery** [comment]
至于混在一起，我看看你的py脚本，你可能需要改一些东西

**Comment by xueli12** [comment]
我用subprocess来监控N_m3u8DL-RE的输出信息时，发现输出不了进度条到窗口中。
``` python
import subprocess
import sys

command = ['N_m3u8DL-RE', 'https://f43.frontgross.com/hl/ta_miau/2024-01-03,06-51/index.m3u8?md5=2j-Va4Jg9bwxAM9sA9LYZg&expires=1707062670&origin=110&mid=9fcba76f-d852-4e58-83a2-ffb8e653682e&akey=6c9c01ce-07af-45c2-afb3-b60cf0017bcb&p=6002&request_id=', '-H', 'User-Agent:Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36', '--force-ansi-console', '--download-retry-count', '99', '--thread-count', '8']

# run the command and capture the output in real time
last_line = ""
progress_info = ""
with subprocess.Popen(command, stdout=subprocess.PIPE, bufsize=1, universal_newlines=True, encoding='utf-8-sig') as p:
    while p.poll() is None:  # while the process is still running
        line = p.stdout.readline().strip()  # read the next line
        if "Progress:" in line:
            progress_info = line  # capture the progress information
        last_line = line  # capture the last line
        print(line)  # print the line
        if "One or more errors occurred" in line:
            print("请重新下载")
            break

# print the progress and last line after the process ends
print("进度信息: ", progress_info)
print("最后一行: ", last_line) 
```

**Comment by RikaCelery** [comment]
![image](https://github.com/RikaCelery/N_m3u8DL-RE/assets/94585272/fe122347-4b68-4958-ae4a-2c6d7b2dd5a5)

```python
import json
import string
import subprocess
import sys

command = ['N_m3u8DL-RE', 'https://f43.frontgross.com/hl/ta_miau/2024-01-03,06-51/index.m3u8?md5=2j-Va4Jg9bwxAM9sA9LYZg&expires=1707062670&origin=110&mid=9fcba76f-d852-4e58-83a2-ffb8e653682e&akey=6c9c01ce-07af-45c2-afb3-b60cf0017bcb&p=6002&request_id=', '-H', 'User-Agent:Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36', '--force-ansi-console', '--download-retry-count', '99', '--thread-count', '8']

# run the command and capture the output in real time
last_line = ""
progress_info = ""
with subprocess.Popen(command, stdout=subprocess.PIPE, bufsize=1, universal_newlines=True, encoding='gbk') as p:
    while p.poll() is None:  # while the process is still running
        line = p.stdout.readline().strip()  # read the next line
        if line.isspace():
            continue
        if "%" in line:
            print(line)  # print the line
            progress_info = line  # capture the progress information
        last_line = line  # capture the last line
        if "One or more errors occurred" in line:
            print("请重新下载")
            break

# print the progress and last line after the process ends
print("进度信息: ", progress_info)
print("最后一行: ", last_line) 
```
用这个

**Comment by RikaCelery** [comment]
输出不了好像是python自己解码错误，多试试不同的编码，进度百分比可以用正则表达式获取，然后你批量调用的话输出每个任务的百分比就行了（输出完成/未完成也可以）

**Comment by xueli12** [comment]
好的 我再试试 太感谢大佬了

**Comment by RikaCelery** [comment]
爬的啥啊stripchat吗，我之前写过一个录制监听+ffmpeg自动转码hevc的东西hhh

一开始用的这个但是不是怎么好用后来就自己写下载器了

**Comment by xueli12** [comment]
> 爬的啥啊stripchat吗，我之前写过一个录制监听+ffmpeg自动转码hevc的东西hhh

下载recu.me的视频

**Comment by RikaCelery** [comment]
哦这样，我推荐你看看[stash](https://github.com/stashapp/stash/)这个项目，管理起来比较方便

**Comment by xueli12** [comment]
> 爬的啥啊stripchat吗，我之前写过一个录制监听+ffmpeg自动转码hevc的东西hhh

牛啊，我还不会自己写，毕竟我还只是入门阶段。我是用别人的程序ctbrec 录stripchat这个站

**Comment by xueli12** [comment]
> 哦这样，我推荐你看看[stash](https://github.com/stashapp/stash/)这个项目，管理起来比较方便

感谢推荐

**Comment by xueli12** [comment]
> 哦这样，我推荐你看看[stash](https://github.com/stashapp/stash/)这个项目，管理起来比较方便
大佬，请教下这个stash工具。我发现添加了recu视频的网站后，程序没法显示信息，应该怎么让stash去解析和下载呀
![image](https://github.com/RikaCelery/N_m3u8DL-RE/assets/51438016/c547d0e2-54d9-46b5-a0d9-e6ccc352dd41)

**Comment by RikaCelery** [comment]
> > 哦这样，我推荐你看看[stash](https://github.com/stashapp/stash/)这个项目，管理起来比较方便
> 大佬，请教下这个stash工具。我发现添加了recu视频的网站后，程序没法显示信息，应该怎么让stash去解析和下载呀
> ![image](https://github.com/RikaCelery/N_m3u8DL-RE/assets/51438016/c547d0e2-54d9-46b5-a0d9-e6ccc352dd41)
> 
~~我还没有使用stash~~
大概是官方没有支持需要自己写解析
也可能是stash就是个管理软件不是全自动爬虫

**Comment by xueli12** [comment]
大佬，我开了几十个线程同时下载，但是当分片下完后，Nm3u8dl会调用ffmpeg合并，然后下载也变慢了，因为很多个线程同时合并，硬盘也卡卡的。有没有什么办法只允许几个Nm3u8dl调用ffmpeg合并，让其他线程的Nm3u8dl调用ffmpeg合并的命令排在队列中呀？

**Comment by RikaCelery** [comment]
> 大佬，我开了几十个线程同时下载，但是当分片下完后，Nm3u8dl会调用ffmpeg合并，然后下载也变慢了，因为很多个线程同时合并，硬盘也卡卡的。有没有什么办法只允许几个Nm3u8dl调用ffmpeg合并，让其他线程的Nm3u8dl调用ffmpeg合并的命令排在队列中呀？

试试挂起线程

**Comment by xueli12** [comment]
> 爬的啥啊stripchat吗，我之前写过一个录制监听+ffmpeg自动转码hevc的东西hhh
> 
> 一开始用的这个但是不是怎么好用后来就自己写下载器了

大佬，你所使用的ffmpeg录制直播流的命令可否分享一下？感谢

**Comment by RikaCelery** [comment]
> > 爬的啥啊stripchat吗，我之前写过一个录制监听+ffmpeg自动转码hevc的东西hhh
> > 一开始用的这个但是不是怎么好用后来就自己写下载器了
> 
> 大佬，你所使用的ffmpeg录制直播流的命令可否分享一下？感谢

不是ffmpeg，单独的小软件，录制结束后用ffmpeg过一遍修复一下错误的时间轴

**Comment by xueli12** [comment]
> > > 爬的啥啊stripchat吗，我之前写过一个录制监听+ffmpeg自动转码hevc的东西hhh
> > > 一开始用的这个但是不是怎么好用后来就自己写下载器了
> > 
> > 
> > 大佬，你所使用的ffmpeg录制直播流的命令可否分享一下？感谢
> 
> 不是ffmpeg，单独的小软件，录制结束后用ffmpeg过一遍修复一下错误的时间轴

大佬我还想请教一个问题：

是关于这个站的https://www.popkontv.com/

ponkontv在获取直播流的请求头中需要加入authorization，好像是要在用账户密码登录后才会发放这个获取直播流的authorization，我用cookie方式登录不知道如何获取这个。于是后面我采用登录的方式获取authorization，

但是我在登录时遇到一个问题，在登录时会有如下面的一个base 开头的authorization，这个授权头貌似是js生成的，我换了几个浏览器和换了不同电脑 这个base开头的授权头都是一样，不知道是如何生成的，虽说现在可以用这个base开头的授权头来登录，但是万一哪天变化了，因此想请教一下，貌似是js生成的。

![20240223022651](https://github.com/ihmily/DouyinLiveRecorder/assets/51438016/c0c64ae4-5783-4510-90bd-99b36caafb80)

![ZE(JU_Q_SSDRTQDS_YCHH%1](https://github.com/RikaCelery/N_m3u8DL-RE/assets/51438016/b2068427-7fe7-41dd-b62d-75bdf3169b1e)

**Comment by RikaCelery** [comment]
你可以把他扔进node里面跑，混淆完了代码基本不能看了

**Comment by Ephemeralwanning** [comment]
![image](https://github.com/user-attachments/assets/47827c7f-bb9c-4850-aa9e-0b009e7b465b)
大佬们如果链接下载的有多部分内容，进度条返回的咋把它们分开处理呀（不知道咋样才能默认选择下载视频，原来命令行只有个"--auto-select"参数）

**Reply by RikaCelery** [reply]
我没有听懂你的意思，进度条返回这个需要你自己想办法解析
--auto-select是自动选择最高清晰度视频下载

**Comment by Ephemeralwanning** [comment]
```
21:09:40.700 INFO : Vid *CENC 1920x1080 | 5723 Kbps | p0va0br5723277 | avc1.4d4028 | 1761 Segments | ~58m40s
21:09:40.700 INFO : Vid *CENC 1280x720 | 3731 Kbps | p0va0br3731517 | avc1.4d4028 | 1761 Segments | ~58m40s
21:09:40.700 INFO : Vid *CENC 720x404 | 2861 Kbps | p0va0br2861150 | avc1.4d401e | 1761 Segments | ~58m40s
21:09:40.701 INFO : Vid *CENC 720x404 | 1865 Kbps | p0va0br1865421 | avc1.4d401e | 1761 Segments | ~58m40s
21:09:40.701 INFO : Vid *CENC 560x316 | 901 Kbps | p0va0br901180 | avc1.4d401e | 1761 Segments | ~58m40s
21:09:40.701 INFO : Vid *CENC 320x180 | 267 Kbps | p0va0br267462 | avc1.42c00d | 1761 Segments | ~58m40s
21:09:40.702 INFO : Aud *CENC p0aa0br256000 | 256 Kbps | mp4a.40.2 | eng | 2CH | 1761 Segments | ~58m40s
21:09:40.704 ERROR: Cannot see if a key has been pressed when either application does not have a console or when console input has been redirected from a file. Try Console.In.Peek.
```
如果按照上面不加"--auto-select"就不会下载 加了就默认下载最大的

**Reply by RikaCelery** [reply]
不加会等你去选择下载哪一个，加了是自动下载，你可以用正则选择指定的视频流
`--select-video <OPTIONS>`
```
通过正则表达式选择符合要求的视频流. 你能够以:分隔形式指定如下参数:

id=REGEX:lang=REGEX:name=REGEX:codec=REGEX:res=REGEX:frame=REGEX
segsMin=number:segsMax=number:ch=REGEX:range=REGEX:url=REGEX
plistDurMin=hms:plistDurMax=hms:role=string:for=FOR

* for=FOR: 选择方式. best[number], worst[number], all (默认: best)

例如:
# 选择最佳视频
-sv best
# 选择4K+HEVC视频
-sv res="3840*":codec=hvc1:for=best
# 选择长度大于1小时20分钟30秒的视频
-sv plistDurMin="1h20m30s":for=best
-sv role="main":for:best
```

**Reply by Ephemeralwanning** [reply]
> 不加会等你去选择下载哪一个，加了是自动下载，你可以用正则选择指定的视频流 `--select-video <OPTIONS>`
> 
> ```
> 通过正则表达式选择符合要求的视频流. 你能够以:分隔形式指定如下参数:
> 
> id=REGEX:lang=REGEX:name=REGEX:codec=REGEX:res=REGEX:frame=REGEX
> segsMin=number:segsMax=number:ch=REGEX:range=REGEX:url=REGEX
> plistDurMin=hms:plistDurMax=hms:role=string:for=FOR
> 
> * for=FOR: 选择方式. best[number], worst[number], all (默认: best)
> 
> 例如:
> # 选择最佳视频
> -sv best
> # 选择4K+HEVC视频
> -sv res="3840*":codec=hvc1:for=best
> # 选择长度大于1小时20分钟30秒的视频
> -sv plistDurMin="1h20m30s":for=best
> -sv role="main":for:best
> ```

我利用上面的代码将输出截取到python不加"--auto-select"参数程序进程就会自己退出，没办法根据返回的结果进行交互，大佬我就是想不明白这点

**Reply by RikaCelery** [reply]
在子进程运行的时候和你用shell运行的环境不一样，没法正常和程序交互
你要是想用选择程序流的话我建议是python先解析一下流列表，再吧想要的流扔给N_m3u8DL下载

**Reply by Ephemeralwanning** [reply]
> 在子进程运行的时候和你用shell运行的环境不一样，没法正常和程序交互 你要是想用选择程序流的话我建议是python先解析一下流列表，再吧想要的流扔给N_m3u8DL下载

好的谢谢大佬 我试试
